File: search-a-2d-matrix.cpp
class Solution {

public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int size = matrix.size();
        if (size == 0) return false;
        int colSz = matrix[0].size(), col = colSz-1, row = 0;
        while (row < size && col >= 0) {
            if (matrix[row][col] == target) return true;
            if (matrix[row][col] < target) {
                row++;
            } else {
                col--;
            }
        }
        return false;
    }
};

File: search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) return false;
        int m = matrix.length, n = matrix[0].length;
        int start = 0, end = m * n - 1;
        
        while (start < end) {
            int mid = (start + end) / 2;
            int i = mid / n, j = mid % n;
            if (target > matrix[i][j]) {
                start = mid+1;
            } else if (target < matrix[i][j]) {
                end = mid;
            } else {
                return true;
            }
        }
        return matrix[end/n][end%n] == target;
    }
}
